File: sdk/python/examples/cron_example.py
"""
Cron Scheduling Example

Demonstrates how to manage scheduled tasks via the cron system:
- Creating cron jobs with standard cron syntax
- Parsing natural language schedules
- Job lifecycle (pause, resume, update, delete)
- Manual job execution
- Viewing execution history
"""

import asyncio
from a3s_code import A3sClient


async def cron_example():
    print("=" * 60)
    print("Cron Scheduling Example")
    print("=" * 60)
    print()

    async with A3sClient(address="localhost:4088") as client:
        # =====================================================================
        # Parse Natural Language Schedule
        # =====================================================================
        print("1. Parsing natural language schedules...")
        schedules = [
            "every 5 minutes",
            "every day at 2am",
            "every Monday at 9:30",
        ]
        for text in schedules:
            result = await client.parse_cron_schedule(text)
            cron_expr = result.get("expression", "N/A")
            description = result.get("description", "")
            print(f"  '{text}' → {cron_expr}  ({description})")
        print()

        # =====================================================================
        # Create Cron Jobs
        # =====================================================================
        print("2. Creating cron jobs...")

        # Job 1: Run every 5 minutes
        job1 = await client.create_cron_job(
            name="health-check",
            schedule="*/5 * * * *",
            command="curl -s http://localhost:8080/health",
            timeout_ms=10000,
        )
        job1_id = job1.get("id", "")
        print(f"✓ Created '{job1.get('name')}': {job1.get('schedule')} (id: {job1_id})")

        # Job 2: Daily backup at 2am
        job2 = await client.create_cron_job(
            name="daily-backup",
            schedule="0 2 * * *",
            command="tar -czf /tmp/backup-$(date +%Y%m%d).tar.gz /data",
            timeout_ms=300000,
        )
        job2_id = job2.get("id", "")
        print(f"✓ Created '{job2.get('name')}': {job2.get('schedule')} (id: {job2_id})")

        # Job 3: Weekly report
        job3 = await client.create_cron_job(
            name="weekly-report",
            schedule="0 9 * * 1",
            command="python /scripts/generate_report.py",
        )
        job3_id = job3.get("id", "")
        print(f"✓ Created '{job3.get('name')}': {job3.get('schedule')} (id: {job3_id})")
        print()

        # =====================================================================
        # List All Jobs
        # =====================================================================
        print("3. Listing all cron jobs...")
        jobs = await client.list_cron_jobs()
        print(f"✓ Found {len(jobs)} jobs:")
        for job in jobs:
            print(f"  - {job.name} [{job.status.name}]: {job.schedule}")
            print(f"    Command: {job.command[:60]}...")
            print(f"    Runs: {job.run_count} success, {job.fail_count} failed")
        print()

        # =====================================================================
        # Get Job Details
        # =====================================================================
        print("4. Getting job details...")

        # By ID
        job_detail = await client.get_cron_job(id=job1_id)
        if job_detail:
            print(f"✓ Job by ID: {job_detail.name}")
            print(f"  Schedule: {job_detail.schedule}")
            print(f"  Timeout: {job_detail.timeout_ms}ms")
            if job_detail.next_run:
                print(f"  Next run: {job_detail.next_run}")

        # By name
        job_by_name = await client.get_cron_job(name="daily-backup")
        if job_by_name:
            print(f"✓ Job by name: {job_by_name.name} (id: {job_by_name.id})")
        print()

        # =====================================================================
        # Update a Job
        # =====================================================================
        print("5. Updating job schedule...")
        updated = await client.update_cron_job(
            id=job1_id,
            schedule="*/10 * * * *",  # Change to every 10 minutes
            timeout_ms=15000,
        )
        print(f"✓ Updated '{updated.get('name')}': schedule → {updated.get('schedule')}")
        print()

        # =====================================================================
        # Pause and Resume
        # =====================================================================
        print("6. Pausing job...")
        pause_result = await client.pause_cron_job(job2_id)
        print(f"✓ Paused: {pause_result.get('name')} → status: {pause_result.get('status')}")

        print("   Resuming job...")
        resume_result = await client.resume_cron_job(job2_id)
        print(f"✓ Resumed: {resume_result.get('name')} → status: {resume_result.get('status')}")
        print()

        # =====================================================================
        # Manual Execution
        # =====================================================================
        print("7. Running job manually...")
        run_result = await client.run_cron_job(job1_id)
        print(f"✓ Execution result:")
        print(f"  Status: {run_result.get('status')}")
        print(f"  Exit code: {run_result.get('exit_code')}")
        stdout = run_result.get("stdout", "")
        if stdout:
            print(f"  Output: {stdout[:200]}")
        stderr = run_result.get("stderr", "")
        if stderr:
            print(f"  Stderr: {stderr[:200]}")
        print()

        # =====================================================================
        # Execution History
        # =====================================================================
        print("8. Getting execution history...")
        history = await client.get_cron_history(job1_id, limit=5)
        print(f"✓ Found {len(history)} executions:")
        for exec_record in history:
            duration = f"{exec_record.duration_ms}ms" if exec_record.duration_ms else "N/A"
            print(f"  - [{exec_record.status.name}] duration={duration}, exit={exec_record.exit_code}")
            if exec_record.stdout:
                print(f"    stdout: {exec_record.stdout[:80]}")
        print()

        # =====================================================================
        # Delete a Job
        # =====================================================================
        print("9. Deleting jobs...")
        for jid in [job1_id, job2_id, job3_id]:
            result = await client.delete_cron_job(jid)
            print(f"✓ Deleted: {result.get('name', jid)}")

        # Verify deletion
        remaining = await client.list_cron_jobs()
        print(f"  Remaining jobs: {len(remaining)}")
        print()

        print("=" * 60)
        print("Cron example complete! ✓")
        print("=" * 60)


if __name__ == "__main__":
    asyncio.run(cron_example())

File: sdk/typescript/examples/src/cron-example.ts
/**
 * Cron Scheduling Example
 *
 * Demonstrates how to manage scheduled tasks via the cron system:
 * - Creating cron jobs with standard cron syntax
 * - Parsing natural language schedules
 * - Job lifecycle (pause, resume, update, delete)
 * - Manual job execution
 * - Viewing execution history
 */

import { A3sClient } from '@a3s-lab/code';

async function cronExample(): Promise<void> {
  console.log('='.repeat(60));
  console.log('Cron Scheduling Example');
  console.log('='.repeat(60));
  console.log();

  const client = new A3sClient({
    address: process.env.A3S_ADDRESS || 'localhost:4088',
  });

  try {
    // Parse natural language schedules
    console.log('1. Parsing natural language schedules...');
    const schedules = ['every 5 minutes', 'every day at 2am', 'every Monday at 9:30'];
    for (const text of schedules) {
      const result = await client.parseCronSchedule(text);
      console.log(`  '${text}' → ${result.cronExpression}  (${result.description})`);
    }
    console.log();

    // Create cron jobs
    console.log('2. Creating cron jobs...');

    const job1 = await client.createCronJob(
      'health-check',
      '*/5 * * * *',
      'curl -s http://localhost:8080/health',
      10000,
    );
    const job1Id = job1.id || '';
    console.log(`✓ Created '${job1.name}': ${job1.schedule} (id: ${job1Id})`);

    const job2 = await client.createCronJob(
      'daily-backup',
      '0 2 * * *',
      'tar -czf /tmp/backup-$(date +%Y%m%d).tar.gz /data',
      300000,
    );
    const job2Id = job2.id || '';
    console.log(`✓ Created '${job2.name}': ${job2.schedule} (id: ${job2Id})`);

    const job3 = await client.createCronJob(
      'weekly-report',
      '0 9 * * 1',
      'python /scripts/generate_report.py',
    );
    const job3Id = job3.id || '';
    console.log(`✓ Created '${job3.name}': ${job3.schedule} (id: ${job3Id})`);
    console.log();

    // List all jobs
    console.log('3. Listing all cron jobs...');
    const jobs = await client.listCronJobs();
    console.log(`✓ Found ${jobs.jobs?.length || 0} jobs:`);
    for (const job of jobs.jobs || []) {
      console.log(`  - ${job.name} [${job.status}]: ${job.schedule}`);
      console.log(`    Command: ${job.command.substring(0, 60)}...`);
    }
    console.log();

    // Get job details
    console.log('4. Getting job details...');
    const jobDetail = await client.getCronJob(job1Id);
    if (jobDetail.job) {
      console.log(`✓ Job by ID: ${jobDetail.job.name}`);
      console.log(`  Schedule: ${jobDetail.job.schedule}`);
      console.log(`  Timeout: ${jobDetail.job.timeoutMs}ms`);
    }

    const jobByName = await client.getCronJob(undefined, 'daily-backup');
    if (jobByName.job) {
      console.log(`✓ Job by name: ${jobByName.job.name} (id: ${jobByName.job.id})`);
    }
    console.log();

    // Update a job
    console.log('5. Updating job schedule...');
    const updated = await client.updateCronJob(job1Id, '*/10 * * * *', undefined, 15000);
    console.log(`✓ Updated: schedule → ${updated.schedule}`);
    console.log();

    // Pause and resume
    console.log('6. Pausing job...');
    const pauseResult = await client.pauseCronJob(job2Id);
    console.log(`✓ Paused: ${pauseResult.name} → status: ${pauseResult.status}`);

    console.log('   Resuming job...');
    const resumeResult = await client.resumeCronJob(job2Id);
    console.log(`✓ Resumed: ${resumeResult.name} → status: ${resumeResult.status}`);
    console.log();

    // Manual execution
    console.log('7. Running job manually...');
    const runResult = await client.runCronJob(job1Id);
    console.log('✓ Execution result:');
    console.log(`  Status: ${runResult.status}`);
    console.log(`  Exit code: ${runResult.exitCode}`);
    if (runResult.stdout) {
      console.log(`  Output: ${runResult.stdout.substring(0, 200)}`);
    }
    console.log();

    // Execution history
    console.log('8. Getting execution history...');
    const history = await client.getCronHistory(job1Id, 5);
    console.log(`✓ Found ${history.executions?.length || 0} executions:`);
    for (const exec of history.executions || []) {
      const duration = exec.durationMs ? `${exec.durationMs}ms` : 'N/A';
      console.log(`  - [${exec.status}] duration=${duration}, exit=${exec.exitCode}`);
    }
    console.log();

    // Delete jobs
    console.log('9. Deleting jobs...');
    for (const jid of [job1Id, job2Id, job3Id]) {
      const result = await client.deleteCronJob(jid);
      console.log(`✓ Deleted: ${result.name || jid}`);
    }

    const remaining = await client.listCronJobs();
    console.log(`  Remaining jobs: ${remaining.jobs?.length || 0}`);
    console.log();

    console.log('='.repeat(60));
    console.log('Cron example complete! ✓');
    console.log('='.repeat(60));
  } catch (error) {
    console.error('Error:', error);
  }
}

cronExample();
